pyrefly/lib/alt/class/dataclass.rs

```diff
@@ -180,6 +187,314 @@ impl<'a, Ans: LookupAnswer> AnswersSolver<'a, Ans> {
         Some(ClassSynthesizedFields::new(fields))
     }
 
+    pub fn call_dataclasses_replace(
+        &self,
+        args: &[Expr],
+        keywords: &[Keyword],
+        range: TextRange,
+        errors: &ErrorCollector,
+    ) -> Type {
+        let call_args = args.map(CallArg::expr_maybe_starred);
+        let call_keywords = keywords.map(CallKeyword::new);
+        let obj_name = Name::new_static("obj");
+
+        let obj_ty = match args.first() {
+            Some(first) => self.expr_infer(first, errors),
+            None => Type::Any(AnyStyle::Implicit),
+        };
+        let mut dataclass_types = SmallSet::new();
+        let mut saw_any = false;
+        let mut saw_non_dataclass = false;
+        self.distribute_over_union(&obj_ty, |ty| {
+            match ty {
+                Type::Any(_) => saw_any = true,
+                Type::ClassType(cls) => {
+                    let cls_metadata = self.get_metadata_for_class(cls.class_object());
+                    let allow_dataclass = cls_metadata.dataclass_metadata().is_some_and(|m| {
+                        m.field_specifiers.iter().any(|k| {
+                            matches!(
+                                k,
+                                CalleeKind::Function(FunctionKind::DataclassField)
+                                    | CalleeKind::Class(ClassKind::DataclassField)
+                            )
+                        })
+                    });
+                    if allow_dataclass {
+                        dataclass_types.insert(cls.clone());
+                    } else {
+                        saw_non_dataclass = true;
+                    }
+                }
+                _ => saw_non_dataclass = true,
+            }
+            ty.clone()
+        });
+
+        let metadata = FuncMetadata {
+            kind: FunctionKind::DataclassReplace,
+            flags: Default::default(),
+        };
+
+        if dataclass_types.is_empty() {
+            if !saw_any {
+                self.error(
+                    errors,
+                    args.first().map(|x| x.range()).unwrap_or(range),
+                    ErrorInfo::Kind(ErrorKind::InvalidArgument),
+                    "dataclasses.replace() should be called on dataclass instances".to_owned(),
+                );
+            }
+            return self.callable_infer(
```

Comment:
Similar to the comment about the empty `args` case above - I'd prefer falling back to the regular function call logic  in edge cases rather than adding more special behavior.